ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala

```diff
@@ -476,62 +555,123 @@ class WebSocketService(
-            },
-            Future.successful((Source.empty, offPrefix)),
-          )
-
       Source
-        .lazyFutureSource(() =>
-          prefilter.map { case (prefiltered, shiftedPrefix) =>
-            val liveFiltered = contractsService
-              .insertDeleteStepSource(
-                jwt,
-                parties,
-                resolved.toList,
-                shiftedPrefix,
-                Terminates.Never,
-              )
-              .via(convertFilterContracts(fn))
-            (prefiltered.map(StepAndErrors(Seq.empty, _)) ++ liveFiltered)
-              .via(emitOffsetTicksAndFilterOutEmptySteps(shiftedPrefix))
-          }
-        )
+        .lazyFutureSource(() => {
+          acsPred
+            .map(fetchAndPreFilterAcs(_, jwt, parties))
+            .cata(
+              _.map { acsAndLiveMarker =>
+                acsAndLiveMarker.flatMapConcat {
+                  case acs @ StepAndErrors(_, Acs(_)) if acs.nonEmpty =>
+                    Source.single(acs)
+                  case StepAndErrors(_, Acs(_)) =>
+                    Source.empty
+                  case liveBegin @ StepAndErrors(_, LiveBegin(offset)) =>
+                    // Produce the predicate that is going to be applied to the incoming transaction stream
+                    // We need to apply this to the request with all the offsets shifted so that each stream
+                    // can filter out anything from liveStartingOffset to the query-specific offset
+                    val acsEnd = offset.toOption.map(domain.StartingOffset(_))
+                    val shiftedRequest = Q.adjustRequest(acsEnd, request)
+
+                    // Get the earliest available offset from where to start from
+                    val liveStartingOffset = Q.liveStartingOffset(acsEnd, shiftedRequest)
+                    val StreamPredicate(_, _, fn, _) =
+                      Q.predicate(shiftedRequest, resolveTemplateId, lookupType)
+
+                    Source.single(liveBegin) ++ contractsService
+                      .insertDeleteStepSource(
+                        jwt,
+                        parties,
+                        resolved.toList,
+                        liveStartingOffset,
+                        Terminates.Never,
+                      )
+                      .via(convertFilterContracts(fn))
+                      .via(emitOffsetTicksAndFilterOutEmptySteps(liveStartingOffset))
+                  case StepAndErrors(_, Txn(_, _)) =>
+                    throw new IllegalStateException("Transaction in ACS")
+                }
+              }, {
+                // Get the earliest available offset from where to start from
+                val liveStartingOffset = Q.liveStartingOffset(offPrefix, request)
+                val StreamPredicate(_, _, fn, _) =
+                  Q.predicate(request, resolveTemplateId, lookupType)
+
+                Future.successful {
+                  contractsService
```

Comment:
I'm relatively sure we always fetched only the ACS from the database, isn't that the case?

Reply:
Oh I see, this is not the case where we have no offset but it’s the case where we made no acs request because everything had an offset. What is slightly confusing is that `insertDeleteStepSource` will fetch the ACS again if you pass `None`. Afaict we will never actually pass `None` here (if `acsRequest` is None`, `liveStartOffset` must be `Some`) so maybe add an assertion and/or a comment to make that explict.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/9847/commits/bf1c88e6ae28ce0be6a3da245d44ada31c0c8eb4